### ISSUE TITLE
NLB S3 access logs attributes don't apply

### ISSUE THREAD

**Issue by Loucool111** [kubernetes-sigs/aws-load-balancer-controller#2149]
**Describe the bug**
When adding this annotation to a k8s service:

```
service.beta.kubernetes.io/aws-load-balancer-attributes: access_logs.s3.enabled=true,access_logs.s3.bucket=bucket,access_logs.s3.prefix=prefix
```

The NLB is not updated, access logs aren't being enabled.
Looking at the logs, the annotation doesn't seem to take effect:

```
...
"loadBalancerAttributes": [
    {
        "key": "access_logs.s3.enabled",
        "value": "false"
    },
    {
        "key": "access_logs.s3.bucket",
        "value": ""
    },
    {
        "key": "access_logs.s3.prefix",
        "value": ""
    },
    {
        "key": "load_balancing.cross_zone.enabled",
        "value": "false"
    }
]
...
```

kubectl shows that the annotation is properly applied:
```
$ kubectl describe service myservice -n mynamespace
Name:                     [redacted]
Namespace:                [redacted]
Labels:                   app=[redacted]
                          app.kubernetes.io/managed-by=Helm
                          chart=[redacted]
                          heritage=Helm
                          release=[redacted]
Annotations:              meta.helm.sh/release-name: [redacted]
                          meta.helm.sh/release-namespace: default
                          service.beta.kubernetes.io/aws-load-balancer-attributes:
                            access_logs.s3.enabled=true,access_logs.s3.bucket=bucket,access_logs.s3.prefix=prefix
                          service.beta.kubernetes.io/aws-load-balancer-nlb-target-type: instance
                          service.beta.kubernetes.io/aws-load-balancer-scheme: internet-facing
                          service.beta.kubernetes.io/aws-load-balancer-type: external
                          service.beta.kubernetes.io/load-balancer-source-ranges: 0.0.0.0/0
Selector:                 app=[redacted],release=[redacted]
Type:                     LoadBalancer
IP Families:              <none>
IP:                       10.100.249.186
IPs:                      10.100.249.186
LoadBalancer Ingress:     [redacted].elb.eu-central-1.amazonaws.com
Port:                     [redacted] 443/TCP
TargetPort:               443/TCP
NodePort:                 [redacted]  31619/TCP
Endpoints:                [redacted]
Session Affinity:         None
External Traffic Policy:  Cluster
Events:                   <none>
```

Can confirm the s3 bucket has proper access policy since other Application Load Balancer can log into it.

**Steps to reproduce**
See above

**Expected outcome**
NLB s3 access logging should be enabled

**Environment**

* AWS Load Balancer controller version
* {"level":"info","ts":1626959998.6067758,"msg":"version","GitVersion":"v2.2.1","GitCommit":"27803e3f8e3b637873f9bb59c56b78de01f65b79","BuildDate":"2021-06-25T17:18:28+0000"}
* Kubernetes version: 1.20.4
* Using EKS (yes/no), if so version? yes: v1.20.4-eks-6b7464

**Additional Context**:

**Comment by kishorj** [comment]
@Loucool111, could you also attach the model from the load balancer controller logs?

**Comment by Loucool111** [comment]
@kishorj Sure, here is the full json model:

```

{
    "id": "[redacted]/[redacted]",
    "resources":
    {
        "AWS::ElasticLoadBalancingV2::Listener":
        {
            "443":
            {
                "spec":
                {
                    "loadBalancerARN":
                    {
                        "$ref": "#/resources/AWS::ElasticLoadBalancingV2::LoadBalancer/LoadBalancer/status/loadBalancerARN"
                    },
                    "port": 443,
                    "protocol": "TCP",
                    "defaultActions": [
                        {
                            "type": "forward",
                            "forwardConfig":
                            {
                                "targetGroups": [
                                    {
                                        "targetGroupARN":
                                        {
                                            "$ref": "#/resources/AWS::ElasticLoadBalancingV2::TargetGroup/[redacted]/[redacted]:443/status/targetGroupARN"
                                        }
                                    }
                                ]
                            }
                        }
                    ]
                }
            }
        },
        "AWS::ElasticLoadBalancingV2::LoadBalancer":
        {
            "LoadBalancer":
            {
                "spec":
                {
                    "name": "k8s-[redacted]-[redacted]-xxxxxxxxxx",
                    "type": "network",
                    "scheme": "internet-facing",
                    "ipAddressType": "ipv4",
                    "subnetMapping": [
                        {
                            "subnetID": "subnet-xxxxxxxxxx"
                        },
                        {
                            "subnetID": "subnet-xxxxxxxxxx"
                        },
                        {
                            "subnetID": "subnet-xxxxxxxxxx"
                        }
                    ],
                    "loadBalancerAttributes": [
                        {
                            "key": "access_logs.s3.enabled",
                            "value": "false"
                        },
                        {
                            "key": "access_logs.s3.bucket",
                            "value": ""
                        },
                        {
                            "key": "access_logs.s3.prefix",
                            "value": ""
                        },
                        {
                            "key": "load_balancing.cross_zone.enabled",
                            "value": "false"
                        }
                    ]
                }
            }
        },
        "AWS::ElasticLoadBalancingV2::TargetGroup":
        {
            "[redacted]/[redacted]:443":
            {
                "spec":
                {
                    "name": "k8s-[redacted]-[redacted]-xxxxxxxxxx",
                    "targetType": "instance",
                    "port": 31619,
                    "protocol": "TCP",
                    "healthCheckConfig":
                    {
                        "port": "traffic-port",
                        "protocol": "TCP",
                        "intervalSeconds": 10,
                        "healthyThresholdCount": 3,
                        "unhealthyThresholdCount": 3
                    },
                    "targetGroupAttributes": [
                        {
                            "key": "proxy_protocol_v2.enabled",
                            "value": "false"
                        }
                    ]
                }
            }
        },
        "K8S::ElasticLoadBalancingV2::TargetGroupBinding":
        {
            "[redacted]/[redacted]:443":
            {
                "spec":
                {
                    "template":
                    {
                        "metadata":
                        {
                            "name": "k8s-[redacted]-[redacted]-xxxxxxxxxx",
                            "namespace": "[redacted]",
                            "creationTimestamp": null
                        },
                        "spec":
                        {
                            "targetGroupARN":
                            {
                                "$ref": "#/resources/AWS::ElasticLoadBalancingV2::TargetGroup/[redacted]/[redacted]:443/status/targetGroupARN"
                            },
                            "targetType": "instance",
                            "serviceRef":
                            {
                                "name": "[redacted]",
                                "port": 443
                            },
                            "networking":
                            {
                                "ingress": [
                                    {
                                        "from": [
                                            {
                                                "ipBlock":
                                                {
                                                    "cidr": "0.0.0.0/0"
                                                }
                                            }
                                        ],
                                        "ports": [
                                            {
                                                "protocol": "TCP",
                                                "port": 31619
                                            }
                                        ]
                                    }
                                ]
                            }
                        }
                    }
                }
            }
        }
    }
}
```

**Comment by kishorj** [comment]
@Loucool111
The annotation `service.beta.kubernetes.io/aws-load-balancer-attributes` is not supported on v2.2.1 release. It will be available in the upcoming v2.3.0 release. Until then, you will have to specify the following annotations for access logging - 
```
service.beta.kubernetes.io/aws-load-balancer-access-log-enabled
service.beta.kubernetes.io/aws-load-balancer-access-log-s3-bucket-name
service.beta.kubernetes.io/aws-load-balancer-access-log-s3-bucket-prefix
```

Also note that the NLB access logs are created only for TLS listeners [1].

References:
[1] https://docs.aws.amazon.com/elasticloadbalancing/latest/network/load-balancer-access-logs.html

**Comment by Loucool111** [comment]
Okay, I didn't see the TLS requirement for logging! Thanks for pointing it out.

Case is closed.
